electrum/plugins/ledger/ledger.py

```diff
@@ -1,37 +1,42 @@
-from struct import pack, unpack
+# Some parts of this code are adapted from bitcoin-core/HWI:
+# https://github.com/bitcoin-core/HWI/blob/e731395bde13362950e9f13e01689c475545e4dc/hwilib/devices/ledger.py
+
+from abc import ABC, abstractmethod
+import base64
 import hashlib
-import sys
-import traceback
-from typing import Optional, Tuple
+from typing import Dict, List, Optional, Sequence, Tuple
 
-from electrum import ecc
-from electrum import bip32
-from electrum.crypto import hash_160
-from electrum.bitcoin import int_to_hex, var_int, is_segwit_script_type, is_b58_address
+import ledger_bitcoin
+from ledger_bitcoin import WalletPolicy, MultisigWallet, AddressType, Chain
+from ledger_bitcoin.exception.errors import DenyError, NotSupportedError, SecurityStatusNotSatisfiedError
+from ledger_bitcoin.key import KeyOriginInfo
+from ledgercomm.interfaces.hid_device import HID
+
+from electrum import bip32, constants, ecc
+from electrum.base_wizard import ScriptTypeNotSupported
 from electrum.bip32 import BIP32Node, convert_bip32_intpath_to_strpath
+from electrum.bitcoin import EncodeBase58Check, int_to_hex, is_b58_address, is_segwit_script_type, var_int
+from electrum.crypto import hash_160
 from electrum.i18n import _
-from electrum.keystore import Hardware_KeyStore
-from electrum.transaction import Transaction, PartialTransaction, PartialTxInput, PartialTxOutput
-from electrum.wallet import Standard_Wallet
-from electrum.util import bfh, bh2u, versiontuple, UserFacingException
-from electrum.base_wizard import ScriptTypeNotSupported
+from electrum.keystore import AddressIndexGeneric, Hardware_KeyStore
 from electrum.logging import get_logger
-from electrum.plugin import runs_in_hwd_thread, Device
-
-from ..hw_wallet import HW_PluginBase, HardwareClientBase
-from ..hw_wallet.plugin import is_any_tx_output_on_change_branch, validate_op_return_output, LibraryFoundButUnusable
+from electrum.plugin import Device, runs_in_hwd_thread
+from electrum.transaction import PartialTransaction, Transaction
+from electrum.util import bfh, UserFacingException, versiontuple
+from electrum.wallet import Standard_Wallet
 
+from ..hw_wallet import HardwareClientBase, HW_PluginBase
+from ..hw_wallet.plugin import is_any_tx_output_on_change_branch, validate_op_return_output
 
 _logger = get_logger(__name__)
 
-
 try:
     import hid
-    from btchip.btchipComm import HIDDongleHIDAPI, DongleWait
+    from btchip.btchipComm import HIDDongleHIDAPI
     from btchip.btchip import btchip
-    from btchip.btchipUtils import compress_public_key,format_transaction, get_regular_input_script, get_p2sh_input_script
+    from btchip.btchipUtils import compress_public_key
     from btchip.bitcoinTransaction import bitcoinTransaction
-    from btchip.btchipFirmwareWizard import checkFirmware, updateFirmware
+    from btchip.btchipFirmwareWizard import checkFirmware
     from btchip.btchipException import BTChipException
```

Comment:
why can't we just use `ledger_bitcoin.btchip` from the new package? is the old package still needed as a dependency?

Reply:
Indeed, that's a good idea. Done in 4ba45d575fce71fc2f43356b73c676c9500d2196.
I removed the call to `checkFirmware` and the corresponding import, as it's not present in `ledger_bitcoin.btchip`.

Reply:
If we have both `Ledger_Client_Legacy` and `Ledger_Client_New`, we might as well keep the HW.1 support for a bit longer. I think the convenient time to remove it is when removing the legacy protocol.

Also, I don't want to remove support for a device in a patch version, and we don't plan to do a large release atm, so keeping support lets us release this in a patch release.

I've pushed a commit to re-add the HW.1 stuff to your branch: https://github.com/bigspider/electrum/commit/9b82eb6d06f8534652849f3370ab8af7d7ca9c7f, along with a new deprecating warning shown in the UI. We can keep depending on `btchip` while the support is kept. I've tried to separate out the hw.1 specific parts more than before. 
I've also tested signing a few txs with a HW.1.